### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by brunoshine** [NullDesk/TicketDesk#42]
Hi Stephen,

is there any "public" roadmap for this app? I ask this because I've fork'ed it to add some new features and it would be nice to have some sense of what might be overlapping or not.

Thanks

**Comment by StephenRedd** [comment]
When I finally get TD 2.5 to a stable beta state, I plan to spend a bit of time putting together a product site for it --maybe here using github pages. I'll be putting together documentation, roadmaps, etc. 

The list of new user facing features for the TD 2.5 timeframe is really short: Multiple projects, switchable HTML and Markdown text editors, individual control of notification settings, and a follow/watch ticket feature. That's about it.

**Comment by btidoug** [comment]
Is there an option still for AD security? If not, is there a way to still
use it somehow?

**Comment by StephenRedd** [comment]
That depends on exactly how you mean. I have no plans to ever support direct integration with on-prem AD again in the same way that TD 1 and 2 did. It is simply too brittle, too hard to configure, and a huge security liability. But read on... all is not lost.   

TD 2.5 will likely go all the way to release with only local user authentication, with identity managed by the standard aspnet.identity framework. As soon as that's done, I will immediately start a short 2.6 release cycle. The only feature of 2.6 will be a new security system. I've been working on this separately for well over a year. Laterly, .Net authentication, authorization, and identity are a complete mess. 

With the 2.6 security module, Identity will be tracked and stored locally by aspnet.identity, but you will be able to couple it with just about any of the modern authentication middlewares. For those middlewares that supply usable claims and identity data, the system will be able to link the authentication provider's claims to the local aspnet.identity store automatically.  

The goal is to let the admin, with minimal effort, configure ticketdesk to use one or more of the following:

- Azure AD (OpenID Connect, Graph API, WsFederation, ADAL)
- ADFS (oAuth, WsFed, SAML, ADAL)
- Local Users (aspnet.identity)
- Common external providers (oAuth, OpenID Connect, facebook, google, etc.)

So... as long as your AD environment has ADFS, or a third party identity service (e.g. Thinktecture IdentityServer) you'll be able to use it for authentication and identity. If you have federated on-premAD with Azure AD, that will also work just fine too.

The wildcard right now is ADAL, as I've yet to dig into too deeply yet. Much of what the latest releases of ADAL do seems to overlap with the standard OWIN/Katana middlewares for WsFed, OpenID Connect and Graph API.

One scenario I am considering, but am inclined to deliberately avoid, is to let TicketDesk use old windows authentication (via IIS) for the authentication, and link it with aspnet.identity for local profile and role management. This would be almost the same flow as external login providers use. This is also the closest to the old TD 1 and 2 direct AD integration as I am considering. 

The reason I'm leaning away from this is that it imposes a tight dependency on IIS and local windows environments, and is fundamentally incompatible with webapi and *js front-ends (which TD 3 is aiming for). To my knowledge, there are no windows auth owin/katana providers, and even if there were it would almost certainly not work on the next generation ASP.NET 5 and .Net Core frameworks.  

So, yeah... that's why this isn't going to be done in time for TD 2.5. They keep moving my cheese every couple of months, and I'm trying to untangle the mess without coding myself into compabitlity problems with ASP.NET 5 and .Net Core... plus trying to keep it all simple enough to configure without the admin needing a degree in computer security.

**Comment by btidoug** [comment]
Makes sense and I understand.  Just curious as our little install is
running perfectly using the AD integration.  Just trying to avoid multiple
logins for users.

Love what you have done and will download and test all your new work soon!

Thank you!

**Comment by brunoshine** [comment]
When you speak of "Multiple projects" are you thinking in the line of tenants, where "nothing" is shared (i.e. Users, Settings) ?

**Comment by StephenRedd** [comment]
Not for TD 2.5. It'll be the simplest implementation possible --a property (like category) on the ticket to associate it with a project name, and the UI will have a top level filter to scope ticketcenter and new ticket views to just the one project. I dont even plan to have user to project associations in TD 2.5 --it will be purely organizational. It will also be purely optional, and disabled by default... if you aren't using multi-projects it will be hidden from the UI entirely.

At some point between TD 2.5 and 3.0, I plan to revamp roles and permissions --the intent is to allow admins to group staff into teams, users into departments, and that sort of thing. Again, with a lot of design effort aimed at keeping the feature simple to administer, invisible to those that don't use it, and yet still enable useful scenarios for larger organizations. That groundwork is also needed before external support scenarios in TD 3 can be fully addressed. Anyway, when I get to that I'll also expand the concept of projects to be scoped to selected groups and users. Again though, not multi-tenancy in any technical sense of the term. 

That's as ambitious as the open source version of TicketDesk is likely to get.

I did prototype TD with several variations of true multi-tenancy last summer. I took it from row level data isolation to schema isolation, then went all the way up to shared nothing. I have a lot of past experience with multi tenancy. The experiment proved that, even with TicketDesk's modest scope, the effort is a little more trouble than it's worth --at least for the open source version. It adds no value to sites that don't need multi-tenancy, but does add significant complexity across the whole system.

That said though, if you have a little experience with multi-tenant systems design, then TicketDesk isn't very difficult to refactor to support it, especially if you only need to design for one specific degree and type of multi-tenancy that fits your particular organization's needs.
